src/ShellClients/PanelWindow.vala

```diff
@@ -53,13 +41,24 @@ public class Gala.PanelWindow : Object {
 
         clone = new PanelClone (wm, this);
 
-        var monitor_manager = wm.get_display ().get_context ().get_backend ().get_monitor_manager ();
-        monitor_manager.monitors_changed.connect (() => update_anchor (anchor));
-        monitor_manager.monitors_changed_internal.connect (() => update_anchor (anchor));
+        var display = wm.get_display ();
+
+        var monitor_manager = display.get_context ().get_backend ().get_monitor_manager ();
+        monitor_manager.monitors_changed.connect (() => set_hide_mode (clone.hide_mode)); //Make sure barriers are still on the primary monitor
 
-        var workspace_manager = wm.get_display ().get_workspace_manager ();
+        var workspace_manager = display.get_workspace_manager ();
         workspace_manager.workspace_added.connect (update_strut);
         workspace_manager.workspace_removed.connect (update_strut);
+
+        window.size_changed.connect (update_strut);
+        window.position_changed.connect (update_strut);
+
+        window_positioner = new WindowPositioner (display, window, WindowPositioner.Position.from_anchor (anchor));
+
+        notify["anchor"].connect (() => {
```

Comment:
Btw the notify here is also useful because we don't want the handler to run the first time it's set (i.e. with the constructor).